third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -341,22 +341,26 @@ impl<'a> FunctionGenerator<'a> {
     }
 
     /// Balance the stack such that it exactly contains the `result` temps and nothing else. This
-    /// is used for instructions like `return` or `abort` which terminate a block und must leave
-    /// the stack empty at end.
+    /// is used for instructions like `brancg`, `return` or `abort` which terminate a block
+    /// und must leave the stack empty at end.
     fn balance_stack_end_of_block(
         &mut self,
         ctx: &BytecodeContext,
         result: impl AsRef<[TempIndex]>,
     ) {
         let result = result.as_ref();
-        // First ensure the arguments are on the stack.
-        self.abstract_push_args(ctx, result, None);
-        if self.stack.len() != result.len() {
-            // Unfortunately, there is more on the stack than needed.
-            // Need to flush and push again so the stack is empty after return.
+        // If the stack contains already exactly the result and none of the temps is used after,
+        // nothing to do
+        let stack_ready = self.stack == result
+            && self
+                .stack
+                .iter()
+                .enumerate()
+                .all(|(pos, temp)| !ctx.is_alive_after(*temp, &result[pos + 1..], false));
```

Comment:
We already know that `stack == result` here, do we need to include the `remaining_args` here?

If the stack already had the copies of a value, this would make us unnecessarily flush them out and copy back again?

Reply:
Lets better not rely on the invariant that all temps on the stack are distinct. Likely it is true, but I did not want to risk anything here, because and this check is safe if temps are distinct.

If the stack had already true (independent) copies, the temps on it would not be alive-after, so I don't think we flush to much here. I definitely have a test case which shows we need to do the alive-after check. There can be temps on the stack which are not saved yet an are used after, those stem from push_result.

Reply:
Sorry, let me clarify: I was suggesting that we would be okay passing an empty slice to the `remaining_args` of `ctx.is_alive_after`, because we already know that `stack == result`.

If any of the stack values are used after (regardless of whether there are repeated values), then `stack_ready` would still be false.

If none of the stack values are used after, but there were some repeated values, then the stack is still ready - we don't have to do a flush in this particular case (but the current code does it).

Reply:
> If none of the stack values are used after, but there were some repeated values, then the stack is still ready - we don't have to do a flush in this particular case (but the current code does it).

Ok, this may be true... there are two scenarios how temps can end on the abstract stack

- We pushed them as args, in which case we already emitted a copy if needed. In this case we definitely have duplicates on the stack, which a previous bug fix dealt with. But those are copies.
- We pushed them as results of function calls. In this case, there should be no duplicates.

Going for the empty slice then.